`mkDerivation`: support `__functor` functions and clean up

roberth opened the pull request:
- Support `__functor` functions as argument (see commit message for details)
  - Found in https://github.com/NixOS/nixpkgs/pull/299542
- Clean up
- Enable `src`/`version` override warning more broadly (see commit message no. 2 for details)


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
mkDerivation: Call functor when __functor is passed

Considering that no attribute value passed to mkDerivation was
allowed to be a function anyway, this does not cause any changes
for code that used the previous interpretation of __functor,
because that could wouldn't have worked anyway.

In practice, this fixes a problem when an option value from
`lib.types.functionTo` is passed as an argument here.
(`functionTo` returns a `lib.setFunctionArgs` result, which is a
__functor)

roberth committed:
mkDerivation: Un-inline makeDerivationExtensible from makeDerivationExtensibleConst

Considering the code was out of sync, this adds the `version`/`src` warning to
non-"finalAttrs" calls to `mkDerivation`. That didn't work.

(I un-DRY-ed this as a micro-optimization at some point, and I apologize... while
simultaneously blaming contributors who change the file without understanding it,
therefore implicating myself *yet again* and making myself a hypocrite)

roberth committed:
mkDerivation: Inline makeDerivationExtensibleConst

roberth committed:
mkDerivation: Distribute makeDerivationExtensible and if-then-else

roberth committed:
mkDerivation: Reuse toFunction

philiptaron commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -62,14 +62,7 @@ let
-      else
-        (_: fnOrAttrs)
-    );
+  mkDerivation = fnOrAttrs: makeDerivationExtensible (toFunction fnOrAttrs);
```
`toFunction`'s definition is `toFunction = v: if isFunction v then v else k: v;`

✅

philiptaron commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -63,7 +64,7 @@ let
   */
   mkDerivation =
     fnOrAttrs:
-    if builtins.isFunction fnOrAttrs then
```
Feedback for the Nix team: I'd support having `builtins.isFunction` return `true` for callable attrsets, aka those with `__functor`. It's a bug factory having `builtins.isFunction` return `false` but `lib.isFunction` return `true`.

Is there a Nix issue for this? Has this ever been considered? There are obviously backwards compatibility issues, but that's just a path straight to ossification.

As a reviewer, it's hard to determine whether code that used `builtins.isFunction` rather than `lib.isFunction` was an intentional choice or an unintentional one.

roberth replied:
- https://github.com/NixOS/nix/issues/13288

philiptaron commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -136,24 +136,7 @@ let
 
-  #makeDerivationExtensibleConst = attrs: makeDerivationExtensible (_: attrs);
-  # but pre-evaluated for a slight improvement in performance.
-  makeDerivationExtensibleConst =
```
Not a breaking change because it doesn't escape the `let` clause in which it's defined.

Appreciate the mea culpa in the commit, @roberth ❤️

philiptaron reviewed (changes requested):
The changes looks great to me, and the commit separation is greatly appreciated since it made the simplifications obvious.

The swath of eval failures, on the other hand, mean that no good deed is done easily. How should we handle these?

roberth committed:
makeDerivation: Define a version of override to require a prev version

Setting a previously unset `version` attribute is hardly an override,
and more likely just a feature of constructing a package in steps.

roberth committed:
libcollectdclient: Refactor: drop redundant inherit

This triggered a warning. Arguably setting the version to the
same value doesn't require a `src` override (suggestion of the
warning), but let's use this opportunity to just clean it up.

roberth committed:
stdenv/darwin: Ignore version override warning

I don't suppose we need to set `src`, the suggestion of the warning.
This is presumably just to signify that it's not the final build
(not *really*) a version.

roberth committed:
isabelle/polyml: Ignore version override warning for now

roberth committed:
mkDerivation: Add comment about version override check

roberth commented:
> swath of eval failures

They're not failures, but warnings that we don't allow, and that happened because the second commit made the warning apply to all `mkDerivation`s instead of a somewhat arbitrary subset.
This was somewhat expected, and can be handled by just fixing those warnings, similar to how the introduction of the warning was presumably handled in the first place.

philiptaron reviewed (approved):
The version hacks here are nuts. Definitely not blocking on the PR, and thanks for doing them @roberth.
